### PR TITLE
Custom Actions in Python Tutorial

### PR DESCRIPTION
**Pull request by b-butler**
## Description

Adds a tutorial for custom actions written in Python to enable users to take full advantage of this feature.

## Checklist:
- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-examples/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-examples/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of authors](https://github.com/glotzerlab/hoomd-examples/blob/master/AUTHORS.md).

### CODE REVIEW

File: 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb [hunk 1]
```diff
@@ -0,0 +1,395 @@
+{
```
Comment (tommy-waltmann) [inline 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>I think the statement in the constructor, <code>self.energy = energy</code> , is missing an underscore.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

File: 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb [hunk 2]
```diff
@@ -0,0 +1,275 @@
+{
```
Comment (tommy-waltmann) [inline 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>I think in the first line you can remove the comma after considerations.</p><p> </p><p>I'm not quite sure what you mean by "However, first profile and then profile".</p><p> </p><p>Is there a tutorial that is going to be specifically about local snapshots? If not, I think some more explanation about them is warranted in this section.</p><p> </p><p>Rephrase "If accessing state information using local snapshots (...), will typically be faster. They don't copy data or aggregate across MPI ranks." to "Local snapshots (...) provide faster access to the simulation's state information than traditional snapshots because they don't copy data or aggregate across MPI ranks".</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (tommy-waltmann) [inline 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>Maybe say "time complexity" instead of just "complexity".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

**Review by tommy-waltmann — APPROVED** [review]

File: 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb [hunk 3]
```diff
@@ -0,0 +1,107 @@
+{
```
Comment (joaander) [inline 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb:1]: <p>Under "Why Python Custom Actions?" - instead of listing a number of things that we considered as developers, list some concrete things that users can do with them. You can keep the same 4 topics, just reword it to make declarative statements of what a user will be able to do and avoid making any judgements about C++ vs Python or users expertise.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb [hunk 4]
```diff
@@ -0,0 +1,331 @@
+{
```
Comment (joaander) [inline 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>I think it is confusing to declare that "..action will have its <code>act</code> method called &nbsp;it will have an attribute <code>_state</code> accessible..." and the also demonstrate a way to break this.</p><p> </p><p>If you expect users to have to modify <code>attach</code>,  I would use a positive example. For example, demonstrate saving something other than <code>_state</code>  and show that the user should call <code>super().attach(simulation)</code>  because it is responsible for storing <code>_state</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>Do we really need to demonstrate all possible combinations here? In a tutorial format it is often best to show the most common cases and leave the user to read the reference documentation to discover the other permutations. Keep in mind that users <em>will</em> copy and paste this code and expect it to just work.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>I think this would be more readable with one output per code block. Rather than print(property), just have each code block call <code>action.property</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>Do we even need to explain this? Or alternately do we need to show that  user's <code>CustomActions</code></p><p>are themselves loggable? Here we are showing two ways to do the same thing. It would be best to just show the one preferred solution.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>This will be a big abstract to users without a concrete example. Is this really necessary at the tutorial level or could users find it in the reference docs?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

File: 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
```
Comment (joaander) [inline 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>This question seems a bit specific. Maybe something more like "How can I modify the state of the system in a custom updater?"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>And here offer some explanation that this example is a demonstration of how to modify the system state.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>In the earlier sections, you instantiated the class inline. I think it would be more clear to use the same style in all the sections.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

File: 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
```
Comment (joaander) [inline 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>I think these general guidelines are a bit too general to start off with. You talk about the importance of profiling, but don't demonstrate it. The same goes for scipy, cython, numba, etc...</p><p> </p><p>I would invert the order these topics are presented and start by mentioning how the local snapshots avoid copy and gather overheads. Then explain that the local snapshots interoperate through numpy/array interfaces with scipy and other packages in the scientific python ecosytem.</p><p> </p><p>Then, if you want, you can note that higher performance may be achievable with compiled solutions such as numpy, cython, or C++.</p><p> </p><p>If you don't find it too long, you could demonstrate profiling by comparing the run time of simulations using the snapshot based updater vs the local snapshot based one. Are there good profiling tools that work inside jupyter notebooks that would work for profiling HOOMD operations?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>You'll need to explicitly seed the numpy random number generator so that all ranks choose the same particle tag.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

**Comment by joaander** [comment]
@jennyfothergill we would appreciate any suggestions you have to improve this tutorial.

File: 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb (hunk 3 above)
Comment (cbkerr) [reply 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb:1]: <p>+1 to this! I added a similar comment on the index page but it works better here for more specific examples.</p>

Comment (cbkerr) [reply 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb:1]: <p>I find reviewNB a bit clumsy and accidentally added it to the branch instead of the PR, idk if my comment will show up so here it is:</p><blockquote>Could add an example of a concrete benefit here: "For example, if you want to change the particle size during a simulation, you would have to stop the simulation, make the change, and run it again. Now, you can set up a custom action to avoid having to restart the simulation"</blockquote>

Comment (cbkerr) [reply 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb:1]: <p>I am unable to make inline suggestions:</p><ul><li>what is "act"?  <strong>Change? Modify?</strong> Seems a bit odd to define a word with itself.</li><li>What is a hyperparameter? (or link to it if defined in prereqs)</li></ul>

Comment (jennyfothergill) [reply 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb:1]: <p>I also give +1 to concrete examples! It could be good to hint here at what examples will follow.</p>

Comment (jennyfothergill) [reply 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb:1]: <p>New user perspective: I could use a little more explanation of how updaters and tuners are different. I understand that the updater changes the state (could remind users what the state is) and the tuner doesn't.</p><p>But that the tuner "modif(ies) another object's hyperparameters." doesn't help me understand without a little more info.</p><p> </p>

File: 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb (hunk 5 above)
Comment (jennyfothergill) [reply 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>+1 this would help with prior mentioned confusion about tuners vs updaters</p>

File: 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb [hunk 7]
```diff
@@ -0,0 +1,247 @@
+{
```
Comment (jennyfothergill) [inline 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: <p>How does the custom action correctly grab the timestep? You mention </p><p>"(If you are wondering how to access simulation data, there is a mechanism for that which we will go over further down). "</p><p>How would a user know which keywords will work and which wont?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb (hunk 4 above)
Comment (jennyfothergill) [inline 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>typo on line 4 intergrator -&gt; integrator</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (jennyfothergill) [inline 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>I think this example is trying to demonstrate that if the attach method is overwritten the custom action breaks. But it would be confusing to a new user here--especially since the tutorial haven't shown a complex example of a custom action yet.</p><p>It may be useful for explaining how custom actions work under the hood, but I think a user might want to see more complex correct uses first.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb (hunk 5 above)
Comment (jennyfothergill) [inline 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>typo: Lennerd-&gt;Lennard</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

**Comment by jennyfothergill** [comment]
Overall thoughts: 
- I think 01, 02, 04, and 05 are more targeted towards a new user. I would recommend reordering such that these come first and 03 and 06 are last. 
- The motivation behind 03 was a little confusing. What do we want a user to learn? Is it a warning not to overwrite the attach method?
- There is a writer and updater example but no tuner example? Could we add one?
- The custom actions open up so many possibilities! Eventually these tutorials could link to some user submitted examples: dynamic bonding, using ML to tune forcefield, etc.
- I never found a good answer to what variable names/keywords the custom action can access. How could a user know? (e.g. the custom action says timestep and it magically grabs the timestep--how can I query the namespace correctly?)

File: 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb (hunk 5 above)
Comment (cbkerr) [inline 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>What are your thoughts on splitting the "non-trivial" stuff into a new tutorial page? It's weird having to remove the previously-created updater.</p><p><code>sim.operations.updaters.remove(energy_inserter)</code></p><p> </p><p>I see it could also be confusing to have duplicate "boilerplate" stuff between both.</p><p> </p><p>At least I would suggest adding a large heading to the second section.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (cbkerr) [inline 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>I was copying each code block into a single python file to run and it took me a sec to realize this wasn't related to updaters. This is helpful to see but it might be out of scope for this tutorial.  Maybe link to the tutorial on variants instead?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (cbkerr) [reply 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>I missed the very subtle change of </p><p><code>self.energy</code> --&gt; <code>self.energy(timestep)</code> </p><p> </p><p>We should call attention to this somehow. Simplest way could be with a comment on that line or by adding a bolded sentence in the previous paragraph</p>

File: 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb (hunk 4 above)
Comment (b-butler) [reply 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>My concern there is someone will look at the tutorial expect the pressure, and we get an email stating that they want support for having pressure for an action. I wouldn't expect people to read the reference documentation.</p>

File: 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb (hunk 6 above)
Comment (b-butler) [reply 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>We could use snakeviz</p>

Comment (b-butler) [reply 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>Good catch.</p>

File: 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb (hunk 4 above)
Comment (joaander) [reply 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: If we don't expect people to read the reference documentation, then why do we even write it?

Comment (b-butler) [reply 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>I think we need to show this pass through at least. We can make this section smaller.</p>

File: 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb (hunk 5 above)
Comment (b-butler) [reply 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p>We don't currently have one, and I don't know if one would be in the works.</p>

Comment (joaander) [reply 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: The introduction tutorial briefly explains variants: https://hoomd-blue.readthedocs.io/en/latest/tutorial/01-Introducing-Molecular-Dynamics/03-Compressing-the-System.html#Variants

If a link is warranted, use a relative link and sphinx/nbconvert should handle it.

File: 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb (hunk 4 above)
Comment (b-butler) [reply 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>Fair enough.</p>

**Comment by b-butler** [comment]
@jennyfothergill If you would like to give the tutorial another review, I have addressed most of your comments. I did not add a custom tuner because the formula is the same as the other two, users aren't likely to need to create one, and it cuts down on the size.

File: 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
```
Comment (cbkerr) [inline 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb:1]: <p>Calling out the definition of <code>sim.state</code>  in a note feels excessive</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
```
Comment (cbkerr) [inline 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: <p>1 - From how I understand currently, an operation is just an action that is ready to be added (after being wrapped in CustomOperation object), so it's confusing to me to say "custom actions and operations do" here.</p><p>1.1 - Add something saying "a custom operation is a custom action + information about when to use the action in the form of a hoomd.trigger"</p><p> </p><p>2 Suggest re-wording last sentence:</p><p>"To use a custom operation, we must add it to a <code>hoomd.Simulation</code> object."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

File: 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
```
Comment (cbkerr) [inline 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: <p>How is <code>super().attach(simulation)</code>  different from <code>self._state=simulation.state</code>  ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: 03-Custom-Actions-In-Python/05-Custom-Writer.ipynb [hunk 11]
```diff
@@ -0,0 +1,417 @@
+{
```
Comment (cbkerr) [inline 03-Custom-Actions-In-Python/05-Custom-Writer.ipynb:1]: <p>Is there a reason people might want to use h5py over gsd?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
```
Comment (cbkerr) [inline 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>If you demo this, people <em>will</em> use it for profiling even if you tell them they shouldn't.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

File: 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb (hunk 7 above)
Comment (cbkerr) [reply 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: <p>+1 to this confusion</p>

File: 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb (hunk 9 above)
Comment (cbkerr) [inline 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: <p>Suggest moving this to boilerplate code (as in the other notebooks) to avoid interrupting the flow in the description of operations.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Review by cbkerr — COMMENTED** [review]
Overall, I think it's ready to be published. There is a net benefit to having the tutorial out there. I couldn't have figured out how to make a custom action without this tutorial.

However, there are still some remaining confusions, and I had to read the tutorial forwards and backwards to understand some concepts. Those should be addressed in the final version of the tutorial. Many are noted in comments in reviewNB, but I see the following as some of the biggest:

I want to bump @jennyfothergill's comment here since this is really important to allow users to go beyond the content in the tutorial:
> I never found a good answer to what variable names/keywords the custom action can access. How could a user know? (e.g. the custom action says timestep and it magically grabs the timestep--how can I query the namespace correctly?


Also, there is info about operations vs. actions in:
* 02-->"Wrapping Custom Actions"
* 03-->"Custom Action Wrapping"
But I think the idea of actions being a component of operations should be addressed in a subheading of 01. I did not understand this as first presented in section 01: 
```
Looking at HOOMD-blue
from the Python perspective, actions are wrapped
or composed by operations. For example, objects like
`hoomd.update.BoxResize` and `hoomd.hpmc.tune.MoveSize` contain
actions internally.
```

File: 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb (hunk 7 above)
Comment (jennyfothergill) [reply 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: <p>Yeah, I'm still confused on this. Is there a way to list all the variables that can be used in a custom action? Showing the user that could be really helpful.</p>

Comment (b-butler) [reply 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: @cbkerr @jennyfothergill I am somewhat confused by what is meant here. The timestep is passed in as a function argument so the action isn't responsible for querying that. Likewise, the only attribute we add is `_state`. I mention that operations expose their actions attributes which means trigger should not be an attribute name, but their isn't really a list just `_state` and whatever else is added in a custom defined `attach` method.

File: 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb (hunk 10 above)
Comment (b-butler) [reply 03-Custom-Actions-In-Python/03-Custom-Action-Features.ipynb:1]: It isn't but it is more robust since if we add something to the default attach the first won't break.

File: 03-Custom-Actions-In-Python/05-Custom-Writer.ipynb (hunk 11 above)
Comment (b-butler) [reply 03-Custom-Actions-In-Python/05-Custom-Writer.ipynb:1]: Potentially for more intuitive log storage since HDF5 is designed for multidimensional arrays and nested access, but in general no GSD is good at what it does.

File: 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb (hunk 12 above)
Comment (b-butler) [reply 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: Agreed, but this was the easiest way to show performance improvements. I didn't want to look for a package to install, and using `%timeit` on the whole simulation is obviously not representative.

File: 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb (hunk 7 above)
Comment (jennyfothergill) [reply 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: <p>I guess what I'm saying is if the timestep variable in the PrintTimestep class was changed to something nonsensical, would this action still work or does the name matter? Does it work because there is a named property of the Simulation class called timestep? How could a user know what named properties they can use in their custom action? Would <code>dir(sim)</code>  or <code>dir(sim.state)</code>  give them a place to start? Is there a place in the docs we could link?</p>

Comment (joaander) [reply 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: @jennyfothergill @cbkerr The first argument to `act` is the timestep. If you defined your act like this:
```def act(self, data):```
, then the caller will still pass the timestep in the first argument and `data` will be the timestep.

Comment (jennyfothergill) [reply 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: <p>OK, so the variable name in the action doesn't matter, is that right? I still don't understand how it correctly grabs the timestep. It looks to me that the timestep is never passed to the action-- the action is added as a custom writer and then somehow grabs the timestep. How would a user go about printing a value other than timestep?</p><p> </p>

Comment (joaander) [reply 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: @jennyfothergill The C++ code that calls the custom action's act method passes the timestep as a positional argument:
https://github.com/glotzerlab/hoomd-blue/blob/898a8e4b179ffb9a51aa97f0a9ce6b9453191753/hoomd/PythonAnalyzer.cc#L12-L15

This is just the way HOOMD is designed. When you write code that extends another code, you need to follow the same API. There is nothing special about HOOMD in this regard. The API defines the method names that programmers and the rest of the code base expect to exist and the arguments that they take.

File: 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    "`hoomd.Simulation` object. Looking at HOOMD-blue\n",
+    "from the Python perspective, actions are wrapped\n",
+    "or composed by operations. For example, objects like\n",
+    "`hoomd.update.BoxResize` and `hoomd.hpmc.tune.MoveSize` contain\n",
+    "actions internally. In other words, the actions implement the logic \n",
+    "repsonsible for implementing the functionality of an operation\n",
+    "(e.g. resizing the box for `hoomd.update.BoxResize`), while the\n",
+    "operation handles some logistics like determining when the action\n",
+    "will run using a `hoomd.Trigger` object.\n",
+    "\n",
```
Comment (cbkerr) [inline 03-Custom-Actions-In-Python/01-What-are-Actions.ipynb:35]: This rewrite would help resolve my confusion about action vs. operation. I feel like it's more accessible to start with what a user knows already exist in hoomd.

```suggestion
    "Some HOOMD-blue operations you may have already used, like\n",
    "`hoomd.update.BoxResize` or `hoomd.hpmc.tune.MoveSize`, contain\n",
    "actions internally.\n",
    "The actions implement logic \n",
    "responsible for implementing the functionality of the operation\n",
    "(e.g. resizing the box for `hoomd.update.BoxResize`), while the\n",
    "operation handles some logistics like determining when the action\n",
    "will run using a `hoomd.Trigger` object.\n",
    "From the Python perspective, we say that actions are wrapped\n",
    "or composed by operations.\n",
    "\n",
```

File: 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb (hunk 7 above)
Comment (cbkerr) [reply 03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb:1]: It might be enough to link to the API docs here. Once I found [this](https://hoomd-blue.readthedocs.io/en/latest/module-hoomd-custom.html#hoomd.custom.Action.act) it helped.

File: 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb (hunk 12 above)
Comment (joaander) [reply 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>I agree with @cbkerr. If a simple %time magic is enough to demonstrate the performance delta of the whole simulation run, then let's use that. The tutorials should cover simple, reusable (copy and paste enabled) code.</p>

File: 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
```
Comment (joaander) [inline 03-Custom-Actions-In-Python/04-Custom-Updater.ipynb:1]: <p><code>GaussianVariant.__call__</code> should use  the seeded <code>rng</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

File: 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb (hunk 12 above)
Comment (joaander) [inline 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>The Gaussian Variant needs to use the seeded <code>rng</code>  so that the runs are reproducible.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>I find the General Guidelines paragraph awkward, especially the introductory paragraph. I think it could be reworded to be clearer. The whole section may also benefit from being shortened - possibly even to a bulleted list. Start off by stating that users should profile their code to find the bottlenecks, and maybe even link to a good tool to do so. Then transition to the remaining paragraphs by stating that these are common bottlenecks in custom actions and here are ways to improve performance of each bottlenecks. </p><p> </p><p>Something like:</p><p>* <code>State.snapshot</code>  is expensive: Use <code>cpu_local_snapshot</code>  or <code>gpu_local_snapshot</code></p><p> </p><p>Avoid judgemental statements like "easiest and most obvious". </p><p> </p><p>I still don't think that Cython/numba/C extensions are appropriate to mention here. Users that are aware of these methods will be knowledgeable enough to apply them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (joaander) [inline 03-Custom-Actions-In-Python/06-Improving-Performance.ipynb:1]: <p>Define <code>GaussianVariant</code>  only once. Either in the boilerplate code section or here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! This is looking really good now. I have only a few change requests (listed above, posted by reviewnb).

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks, these improvements are great!

I have two final change requests:

1) What is the file `03-Custom-Actions-In-Python/profile-pre-local-snapshot.prof`? It doesn't seem to be referenced in the notebooks, so it should probably be removed.

2) Merge v3-api up into this branch and run the pre-commit configuration added in #22

**Review by joaander — APPROVED** [review]
Thanks!
